File: ui/src/components/midiCatalog.ts
import type { MidiMapping } from '../types/midiMap';

/**
 * Display catalog for the MIDI mapping UI: which targets are mappable and
 * how to present them. The native engine accepts any APVTS parameter id (or
 * block-power id); this list is the UI's curation (setup-domain params like
 * calibration stay out; they describe the rig, not something you perform
 * with).
 */

export interface MappableTarget {
  /** APVTS parameter id, positional block power ("block1Power"), or a
      virtual action id ("presetNext"). */
  id: string;
  name: string;
  /** Section subtitle, mirroring the faceplate's layout. */
  group: string;
  /** Drives the behavior display: continuous knobs are absolute via CC,
      toggles flip on/off, triggers fire an action once per press. */
  kind: 'continuous' | 'toggle' | 'trigger';
}

/** Block-power targets are positional ("Block 1" is a lane's first tone
    block whatever it currently holds), so a mapping survives tone swaps and
    preset loads, like switches on a pedalboard. "blockNPower" is the Left
    lane (the only lane in mono), "rightBlockNPower" the Right lane (stereo
    only; the mapping screen offers them per the live chain). Display-only
    cap (the engine takes up to 64): enough for any realistic pedalboard
    without burying the picker. */
const BLOCK_POWER_TARGETS = 12;

export const MAPPABLE_TARGETS: MappableTarget[] = [
  // Virtual actions (native resolves the ids itself): step through the
  // preset list in browser order, wrapping at the ends, for footswitches
  // programmed with CC / note buttons instead of program changes.
  { id: 'presetPrevious', name: 'Previous Preset', group: 'Presets', kind: 'trigger' },
  { id: 'presetNext', name: 'Next Preset', group: 'Presets', kind: 'trigger' },
  { id: 'inputLevel', name: 'Input Gain', group: 'Global', kind: 'continuous' },
  { id: 'outputLevel', name: 'Output Level', group: 'Global', kind: 'continuous' },
  { id: 'outputBalance', name: 'Output Balance', group: 'Global', kind: 'continuous' },
  { id: 'gateEnabled', name: 'Gate Power', group: 'Noise Gate', kind: 'toggle' },
  { id: 'gateThreshold', name: 'Gate Threshold', group: 'Noise Gate', kind: 'continuous' },
  { id: 'toneEqEnabled', name: 'Tone Stack Power', group: 'Tone Stack', kind: 'toggle' },
  { id: 'toneBass', name: 'Bass', group: 'Tone Stack', kind: 'continuous' },
  { id: 'toneMid', name: 'Mid', group: 'Tone Stack', kind: 'continuous' },
  { id: 'toneTreble', name: 'Treble', group: 'Tone Stack', kind: 'continuous' },
  { id: 'spreadEnabled', name: 'Spread Power', group: 'Spread', kind: 'toggle' },
  { id: 'spreadOffset', name: 'Spread Offset', group: 'Spread', kind: 'continuous' },
  { id: 'spreadWobble', name: 'Spread Wobble', group: 'Spread', kind: 'continuous' },
  // Virtual target like block powers: stereo on/off is chain state, not an
  // APVTS parameter (the native mapper resolves the id itself).
  { id: 'stereoEnabled', name: 'Stereo Mode', group: 'Stereo', kind: 'toggle' },
  { id: 'stereoOffsetEnabled', name: 'Offset Power', group: 'Stereo', kind: 'toggle' },
  { id: 'stereoOffsetTime', name: 'Offset', group: 'Stereo', kind: 'continuous' },
  { id: 'chainPanLeft', name: 'Pan L', group: 'Stereo', kind: 'continuous' },
  { id: 'chainPanRight', name: 'Pan R', group: 'Stereo', kind: 'continuous' },
  ...Array.from({ length: BLOCK_POWER_TARGETS }, (_, i): MappableTarget => ({
    id: `block${i + 1}Power`,
    name: `Block ${i + 1} Power`,
    group: 'Chain',
    kind: 'toggle',
  })),
  ...Array.from({ length: BLOCK_POWER_TARGETS }, (_, i): MappableTarget => ({
    id: `rightBlock${i + 1}Power`,
    name: `Block ${i + 1} Power`,
    group: 'Chain R',
    kind: 'toggle',
  })),
];

export const targetById = new Map(MAPPABLE_TARGETS.map((t) => [t.id, t]));

const NOTE_NAMES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'];

/** 60 → "C4" (scientific pitch, middle C = C4). */
export const midiNoteName = (note: number) =>
  `${NOTE_NAMES[note % 12]}${Math.floor(note / 12) - 1}`;

/** "CC 64" / "Note C2": the mapping row's source column. */
export const sourceLabel = (mapping: MidiMapping) =>
  mapping.source === 'cc' ? `CC ${mapping.number}` : `Note ${midiNoteName(mapping.number)}`;

/** How the pairing behaves, mirroring the native derivation: trigger
    targets fire per press, toggle targets (and any note source) flip
    on/off, everything else tracks the CC value absolutely. */
export const behaviorLabel = (mapping: MidiMapping) => {
  const kind = targetById.get(mapping.targetId)?.kind;
  if (kind === 'trigger') return 'Trigger';
  if (kind === 'toggle' || mapping.source === 'note') return 'Toggle';
  return 'Absolute';
};

File: plugin/ui/core/MidiCatalog.cpp
#include "MidiCatalog.h"

namespace t3k::ui::midi {

const std::vector<MappableTarget>& mappableTargets() {
  using K = TargetKind;
  static const std::vector<MappableTarget> table = [] {
    std::vector<MappableTarget> t = {
        // Virtual actions (native resolves the ids itself): step through the
        // preset list in browser order, wrapping at the ends, for footswitches
        // programmed with CC / note buttons instead of program changes.
        {"presetPrevious", "Previous Preset", "Presets", K::trigger},
        {"presetNext", "Next Preset", "Presets", K::trigger},
        {"inputLevel", "Input Gain", "Global", K::continuous},
        {"outputLevel", "Output Level", "Global", K::continuous},
        {"outputBalance", "Output Balance", "Global", K::continuous},
        {"gateEnabled", "Gate Power", "Noise Gate", K::toggle},
        {"gateThreshold", "Gate Threshold", "Noise Gate", K::continuous},
        {"gateRelease", "Gate Release", "Noise Gate", K::continuous},
        {"gateHold", "Gate Hold", "Noise Gate", K::continuous},
        {"gateRange", "Gate Range", "Noise Gate", K::continuous},
        {"toneEqEnabled", "Tone Stack Power", "Tone Stack", K::toggle},
        {"toneBass", "Bass", "Tone Stack", K::continuous},
        {"toneMid", "Mid", "Tone Stack", K::continuous},
        {"toneTreble", "Treble", "Tone Stack", K::continuous},
        {"spreadEnabled", "Spread Power", "Spread", K::toggle},
        {"spreadOffset", "Spread Offset", "Spread", K::continuous},
        {"spreadWobble", "Spread Wobble", "Spread", K::continuous},
        {"spreadWobbleEnabled", "Spread Wobble Power", "Spread", K::toggle},
        {"spreadCrossover", "Spread Crossover", "Spread", K::continuous},
        {"spreadCrossoverEnabled", "Spread Crossover Power", "Spread", K::toggle},
        {"spreadDiffuseEnabled", "Spread Diffuse Power", "Spread", K::toggle},
        // Virtual target like block powers: stereo on/off is chain state, not
        // an APVTS parameter (the native mapper resolves the id itself).
        {"stereoEnabled", "Stereo Mode", "Stereo", K::toggle},
        {"alignEnabled", "Align Power", "Align", K::toggle},
        {"alignOffset", "Align Offset", "Align", K::continuous},
        {"alignWobble", "Align Wobble", "Align", K::continuous},
        {"alignWobbleEnabled", "Align Wobble Power", "Align", K::toggle},
        {"alignCrossover", "Align Crossover", "Align", K::continuous},
        {"alignCrossoverEnabled", "Align Crossover Power", "Align", K::toggle},
        {"alignDiffuseEnabled", "Align Diffuse Power", "Align", K::toggle},
        {"chainPanLeft", "Pan L", "Stereo", K::continuous},
        {"chainPanRight", "Pan R", "Stereo", K::continuous},
        {"chainSoloLeft", "Solo L", "Stereo", K::toggle},
        {"chainSoloRight", "Solo R", "Stereo", K::toggle},
        {"chainInvertLeft", "Invert L", "Stereo", K::toggle},
        {"chainInvertRight", "Invert R", "Stereo", K::toggle},
    };
    for (int i = 1; i <= kBlockPowerTargets; ++i)
      t.push_back({"block" + juce::String(i) + "Power", "Block " + juce::String(i) + " Power", "Chain", K::toggle});
    for (int i = 1; i <= kBlockPowerTargets; ++i)
      t.push_back({"rightBlock" + juce::String(i) + "Power", "Block " + juce::String(i) + " Power", "Chain R",
                   K::toggle});
    return t;
  }();
  return table;
}

const MappableTarget* targetById(const juce::String& id) {
  for (const auto& t : mappableTargets())
    if (t.id == id) return &t;
  return nullptr;
}

std::optional<BlockPowerTarget> blockPowerTarget(const juce::String& targetId) {
  // ^(right)?[bB]lock(\d+)Power$
  const bool right = targetId.startsWith("right");
  auto rest = right ? targetId.substring(5) : targetId;
  if (!(rest.startsWith("block") || rest.startsWith("Block")) || !rest.endsWith("Power")) return std::nullopt;
  const auto digits = rest.substring(5, rest.length() - 5);
  if (digits.isEmpty() || !digits.containsOnly("0123456789")) return std::nullopt;
  return BlockPowerTarget{digits.getIntValue() - 1, right};
}

juce::String noteName(int note) {
  static const char* const kNames[] = {"C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"};
  return juce::String(kNames[((note % 12) + 12) % 12]) + juce::String(note / 12 - 1);
}

juce::String sourceLabel(const MidiMapping& mapping) {
  return mapping.source == MidiSource::cc ? "CC " + juce::String(mapping.number)
                                          : "Note " + noteName(mapping.number);
}

juce::String behaviorLabel(const MidiMapping& mapping) {
  const auto* target = targetById(mapping.targetId);
  const auto kind = target != nullptr ? std::optional(target->kind) : std::nullopt;
  if (kind == TargetKind::trigger) return "Trigger";
  if (kind == TargetKind::toggle || mapping.source == MidiSource::note) return "Toggle";
  return "Absolute";
}

}  // namespace t3k::ui::midi
